src/test/librbd/test_librbd.cc

```diff
@@ -4232,6 +4232,61 @@ interval_set<uint64_t> round_diff_interval(const interval_set<uint64_t>& diff,
   return rounded_diff;
 }
 
+TEST_F(TestLibRBD, SnapDiff)
+{
+  REQUIRE_FEATURE(RBD_FEATURE_FAST_DIFF);
+
+  rados_ioctx_t ioctx;
+  rados_ioctx_create(_cluster, m_pool_name.c_str(), &ioctx);
+
+  rbd_image_t image;
+  int order = 0;
+  std::string image_name = get_temp_image_name();
+  uint64_t size = 16 << 20;
+  ASSERT_EQ(0, create_image(ioctx, image_name.c_str(), size, &order));
+  ASSERT_EQ(0, rbd_open(ioctx, image_name.c_str(), &image, nullptr));
+
+  char test_data[TEST_IO_SIZE + 1];
+  for (size_t i = 0; i < TEST_IO_SIZE; ++i) {
+    test_data[i] = (char) (rand() % (126 - 33) + 33);
+  }
+  test_data[TEST_IO_SIZE] = '\0';
+
+  ASSERT_PASSED(write_test_data, image, test_data, 0,
+                TEST_IO_SIZE, LIBRADOS_OP_FLAG_FADVISE_NOCACHE);
+
+  interval_set<uint64_t> diff;
+  ASSERT_EQ(0, rbd_diff_iterate2(image, nullptr, 0, size, true, true,
+                                 iterate_cb, &diff));
+  EXPECT_EQ(1 << order, diff.size());
```

Comment:
Argh, this fails with fancy striping enabled (`RBD_FEATURES=127`).  I think you are going to need to get the stripe unit with `rbd_get_stripe_unit(image, &stripe_unit)` and use it instead of `1 << order` here and below.  (By default, stripe unit is equal to object size, that is `1 << order`).

Reply:
No, `stripe_unit * stripe_count` is definitely wrong.  Something else is going on here...

Reply:
I apologize, I mentioned stripe unit without giving it much thought and mislead you.  The actual problem is that your image is too small for the common case logic to work against the `su = 64K, sc = 16` striping pattern that is thrown at the image by the test harness.  Please increase the size of the image to 100M and go back to `1 << order`.

Reply:
To prove to yourself that `stripe_unit * stripe_count` is a bogus value, try increasing the size of the image to something smaller than `object_size * stripe_count` (the cutoff point for the common case logic, 64M in this case).  The expectation as it is now should fail with a value that is bigger than `stripe_unit * stripe_count` (1M in this case) but smaller than `object_size` (4M in this case).